File: static/js/workforce.js
async function wfLoad(view) {
  window._wfView = view;
  spin('wfContent');
  const sl = $('wfSL').value || 4;
  const bud = $('wfBudget').value || 300;
  try {
    if (view === 'full') {
      const d = await api(`/api/workforce/?service_level=${sl}`);
      const shifts = d.shift_summary || [], rows = d.per_time_window || [], costs = d.cost_summary || {};
      const peakN = Math.max(...shifts.map(s => s.recommended_staff));
      set('wfContent', `
        <div class="grid-4">
          <div class="stat"><div class="stat-label">Daily Cost (Peak)</div><div class="stat-value red">€${costs.total_daily_cost_peak_staffing_eur}</div></div>
          <div class="stat"><div class="stat-label">Optimized Cost</div><div class="stat-value green">€${costs.total_daily_cost_optimized_eur}</div></div>
          <div class="stat"><div class="stat-label">Daily Savings</div><div class="stat-value blue">€${costs.total_potential_savings_eur}</div></div>
          <div class="stat"><div class="stat-label">Monthly Savings</div><div class="stat-value purple">€${costs.monthly_savings_if_optimized_eur}</div></div>
        </div>
        <div class="shift-grid">${shifts.map(s => {
          const isPeak = s.recommended_staff === peakN;
          const icons = {'Morning (06–14)':'🌅','Afternoon (14–22)':'☀️','Night (22–06)':'🌙'};
          const colors = {'Morning (06–14)':'var(--amber)','Afternoon (14–22)':'var(--blue)','Night (22–06)':'var(--purple)'};
          const label = s.label||s.shift;
          const col = colors[label]||'var(--green)';
          return `<div class="shift-card ${isPeak?'peak':''}">
            <div class="shift-accent" style="background:${col}"></div>
            <div class="shift-body">
              <div class="shift-header">
                <div>
                  <div class="shift-tag">${label}</div>
                </div>
                <div class="shift-icon">${icons[label]||'⏱'}</div>
              </div>
              <div class="shift-staff">
                <div class="shift-num" style="color:${col}">${s.recommended_staff}</div>
                <div class="shift-unit">staff<br>recommended</div>
              </div>
              <div class="shift-metrics">
                <div class="shift-metric">
                  <span class="shift-metric-label">Avg pts/hr</span>
                  <span class="shift-metric-val" style="color:var(--text)">${s.avg_patients_per_hour}</span>
                </div>
                <div class="shift-metric">
                  <span class="shift-metric-label">Wage/shift</span>
                  <span class="shift-metric-val" style="color:var(--green)">€${s.wage_per_shift_eur}</span>
                </div>
                ${isPeak?`<div class="shift-metric"><span class="shift-metric-label" style="color:var(--green)">Peak shift</span><span style="color:var(--green);font-size:10px">★</span></div>`:''}
              </div>
            </div>
          </div>`;
        }).join('')}
        </div>
        <div class="grid-2">
          <div class="card"><div class="card-title">Demand Curve</div><div class="chart-box" style="height:200px"><canvas id="wfDemChart"></canvas></div></div>
          <div class="card"><div class="card-title">Demand vs Capacity</div><div class="chart-box" style="height:200px"><canvas id="wfCapChart"></canvas></div></div>
        </div>
        <div class="card"><div class="card-title">Per Time Window</div>
          <div class="tbl-wrap"><table>
            <thead><tr><th>Time</th><th>Shift</th><th>Avg Pts/hr</th><th>Staff</th><th>Utilization</th><th>Status</th></tr></thead>
            <tbody>${rows.map(r=>`<tr>
              <td class="val" style="font-size:11px">${r.time_window}</td>
              <td>${r.shift||'—'}</td>
              <td class="val">${r.avg_patients}</td>
              <td class="hi">${r.required_staff}</td>
              <td><div style="display:flex;align-items:center;gap:8px">
                <div style="flex:1;height:5px;background:var(--raised);border-radius:3px;overflow:hidden;min-width:60px">
                  <div style="height:100%;width:${r.utilization_pct}%;background:${r.utilization_pct>85?'var(--amber)':'var(--green)'};border-radius:3px"></div>
                </div>
                <span style="font-family:var(--mono);font-size:10px;color:var(--muted)">${r.utilization_pct}%</span>
              </div></td>
              <td>${r.zero_demand?'<span class="badge badge-dim">No demand</span>':r.understaffed?'<span class="badge badge-red">Alert</span>':'<span class="badge badge-green">OK</span>'}</td>
            </tr>`).join('')}</tbody>
          </table></div>
        </div>`);
      setTimeout(() => {
        const L = rows.map(r=>r.time_window.split(/[-–]/)[0].trim());
        const P = rows.map(r=>r.avg_patients), S = rows.map(r=>r.required_staff*parseInt(sl));
        mkChart('wfDemChart','line',L,[{data:P,borderColor:'#00e5a0',backgroundColor:'rgba(0,229,160,.07)',fill:true,tension:.4,pointRadius:3,pointBackgroundColor:'#00e5a0'}],{plugins:{legend:{display:false}},scales:{x:{ticks:{color:'#6b7280',maxRotation:45,font:{size:9,family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}},y:{beginAtZero:true,ticks:{color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}}}});
        mkChart('wfCapChart','bar',L,[
          {label:'Demand',data:P,backgroundColor:'rgba(56,189,248,.55)',borderRadius:3},
          {label:'Capacity',data:S,backgroundColor:'rgba(0,229,160,.35)',borderRadius:3}
        ],{plugins:{legend:{display:true,position:'top',labels:{color:'#9ca3af',boxWidth:10,font:{family:"'IBM Plex Mono'",size:10}}}},scales:{x:{ticks:{color:'#6b7280',maxRotation:45,font:{size:9,family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}},y:{ticks:{color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}}}});
      },60);

    } else if (view === 'heatmap') {
      const d = await api(`/api/workforce/heatmap?service_level=${sl}`);
      const curve = d.demand_curve || [];
      const maxP = Math.max(...curve.map(r=>r.patients), 1);
      const colorFor = (pts, shift) => {
        const i = pts/maxP;
        if (shift==='Shift 1') return `rgba(0,229,160,${0.15+i*0.7})`;
        if (shift==='Shift 2') return `rgba(56,189,248,${0.15+i*0.7})`;
        return `rgba(167,139,250,${0.15+i*0.7})`;
      };
      set('wfContent', `
        <div class="card">
          <div class="card-title-row">
            <div class="card-title">Patient Demand Heatmap — Hour by Hour</div>
            <div style="display:flex;gap:12px;font-size:11px">
              <span style="color:var(--green)">■ Shift 1</span>
              <span style="color:var(--blue)">■ Shift 2</span>
              <span style="color:var(--purple)">■ Shift 3</span>
            </div>
          </div>
          ${curve.map(r=>`<div class="hm-row">
            <div class="hm-label">${r.time.split(/[-–]/)[0].trim()}</div>
            <div class="hm-bar"><div class="hm-fill" style="width:${r.patients/maxP*100}%;background:${colorFor(r.patients,r.shift)}"></div></div>
            <div class="hm-val">${r.patients} pts · ${r.staff} staff</div>
          </div>`).join('')}
        </div>
        <div class="card"><div class="card-title">Demand Curve</div><div class="chart-box" style="height:200px"><canvas id="wfHmLine"></canvas></div></div>
      `);
      setTimeout(()=>{const L=curve.map(r=>r.time.split(/[-–]/)[0].trim());const P=curve.map(r=>r.patients);mkChart('wfHmLine','line',L,[{data:P,borderColor:'#00e5a0',fill:true,backgroundColor:'rgba(0,229,160,.07)',tension:.4,pointRadius:3,pointBackgroundColor:'#00e5a0'}],{plugins:{legend:{display:false}},scales:{x:{ticks:{color:'#6b7280',maxRotation:45,font:{size:9,family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}},y:{beginAtZero:true,ticks:{color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}}}});},50);

    } else if (view === 'alerts') {
      const d = await api(`/api/workforce/alerts?service_level=${sl}`);
      const under = d.understaffing_alerts||[], over = d.overstaffing_alerts||[], zero = d.zero_demand_windows||[];
      const mkList = (list, type, empty) => !list.length
        ? `<div class="alert alert-green"><span>✓</span><div class="body">${empty}</div></div>`
        : list.map(a=>`<div class="alert alert-${a.severity==='HIGH'?'red':'amber'}">
            <span>${a.severity==='HIGH'?'🔴':'🟡'}</span>
            <div class="body"><strong>${a.time} · ${a.shift||''}</strong><br>${a.action||a.message||''}${a.wasted_cost_eur?`<br>Wasted: €${a.wasted_cost_eur}/hr`:''}
            </div></div>`).join('');
      set('wfContent', `
        <div class="grid-4">
          <div class="stat"><div class="stat-label">Total Alerts</div><div class="stat-value ${d.total_alerts>0?'amber':'green'}">${d.total_alerts}</div></div>
          <div class="stat"><div class="stat-label">Understaffing</div><div class="stat-value red">${under.length}</div></div>
          <div class="stat"><div class="stat-label">Overstaffing</div><div class="stat-value amber">${over.length}</div></div>
          <div class="stat"><div class="stat-label">Zero Demand</div><div class="stat-value blue">${zero.length}</div></div>
        </div>
        <div class="card"><div class="card-title">⚠ Understaffing Alerts</div>${mkList(under,'under','No understaffing detected — coverage is adequate.')}</div>
        <div class="card"><div class="card-title">💸 Overstaffing Alerts</div>${mkList(over,'over','No overstaffing detected.')}</div>
        ${zero.length?`<div class="card"><div class="card-title">○ Zero Demand Windows</div>${zero.map(z=>`<div style="padding:8px 0;border-bottom:1px solid var(--border);font-size:12px;color:var(--muted)">${z.time} · ${z.shift}</div>`).join('')}</div>`:''}`);

    } else if (view === 'optimize') {
      const d = await api(`/api/workforce/optimize?service_level=${sl}&budget_eur=${bud}`);
      const scen = d.scenarios || {};
      set('wfContent', `
        <div class="alert alert-blue"><span>⚡</span><div class="body">Three staffing scenarios per shift — minimum cost, full recommended coverage, and budget-constrained (€${bud}/day).</div></div>
        ${Object.entries(scen).map(([shift, s])=>`
          <div class="card">
            <div class="card-title">${s.label}</div>
            <div class="grid-3">
              <div style="text-align:center;padding:14px;background:var(--raised);border-radius:var(--r2)">
                <div style="font-family:var(--mono);font-size:32px;font-weight:600;color:var(--muted)">${s.minimum_coverage}</div>
                <div style="font-size:11px;color:var(--muted);margin-top:4px">Min coverage<br>€${s.minimum_cost_eur}/shift</div>
              </div>
              <div style="text-align:center;padding:14px;background:rgba(0,229,160,.07);border:1px solid rgba(0,229,160,.2);border-radius:var(--r2)">
                <div style="font-family:var(--mono);font-size:32px;font-weight:600;color:var(--green)">${s.recommended}</div>
                <div style="font-size:11px;color:var(--muted);margin-top:4px">Recommended<br>€${s.full_coverage_cost_eur}/shift</div>
              </div>
              <div style="text-align:center;padding:14px;background:rgba(245,158,11,.07);border:1px solid rgba(245,158,11,.2);border-radius:var(--r2)">
                <div style="font-family:var(--mono);font-size:32px;font-weight:600;color:var(--amber)">${s.budget_constrained_staff}</div>
                <div style="font-size:11px;color:var(--muted);margin-top:4px">Budget (€${bud})<br>€${s.budget_cost_eur}/shift</div>
              </div>
            </div>
            <div style="margin-top:12px;font-size:12px;color:var(--muted)">${s.recommendation_reason}</div>
          </div>`).join('')}`);
    }
  } catch(e) { oops('wfContent', e); }
}

File: routers/workforce.py
"""
Workforce Planning Router
==========================
Endpoints:
  GET /               - Full shift analysis: required staff, alerts, costs
  GET /optimize       - Shift optimization scenarios
  GET /alerts         - Understaffing / overstaffing alerts
  GET /heatmap        - Hour x demand matrix for heatmap
  GET /raw            - Raw dataset

Dataset: fau_medical_staff.csv
"""
import math
import pandas as pd
import numpy as np
from fastapi import APIRouter, Query
from .utils import load_csv, df_info

router = APIRouter()
DEFAULT_FILE = "fau_medical_staff.csv"
DEFAULT_SL   = 4   # patients per staff per hour
WAGES        = {"Shift 1": 45, "Shift 2": 45, "Shift 3": 60}   # EUR per 8h shift
SHIFT_NAMES  = {"Shift 1": "Morning (06–14)", "Shift 2": "Afternoon (14–22)", "Shift 3": "Night (22–06)"}


def _load_clean(filename: str):
    df = load_csv(filename)
    df.columns = [c.strip() for c in df.columns]
    # Drop wage row
    mask = df["Time Windows"].astype(str).str.contains("Wage", case=False, na=False)
    wages_row = df[mask]
    df = df[~mask].copy()
    df["Avg_Patient_Number"] = pd.to_numeric(df["Avg_Patient_Number"], errors="coerce")

    # Parse actual wages from dataset if present
    extracted = {}
    if not wages_row.empty:
        r = wages_row.iloc[0]
        for s in ["Shift 1", "Shift 2", "Shift 3"]:
            try:
                extracted[s] = float(r[s])
            except Exception:
                pass
    wages = {**WAGES, **extracted}

    # Detect shift per row
    rows = []
    for _, r in df.iterrows():
        shift = None
        for s in ["Shift 1", "Shift 2", "Shift 3"]:
            if s in r and pd.notna(r[s]) and str(r[s]).strip() != "":
                shift = s
                break
        rows.append({
            "time_window":    str(r["Time Windows"]).strip(),
            "shift":          shift,
            "avg_patients":   float(r["Avg_Patient_Number"]) if pd.notna(r["Avg_Patient_Number"]) else 0.0,
        })
    return pd.DataFrame(rows), wages


def _compute(row_df: pd.DataFrame, sl: int, wages: dict):
    rows = []
    for _, r in row_df.iterrows():
        avg = r["avg_patients"]
        required = math.ceil(avg / sl) if avg > 0 else 0
        capacity = required * sl
        util = round(avg / capacity * 100, 1) if capacity > 0 else 0.0
        rows.append({
            "time_window":    r["time_window"],
            "shift":          r["shift"],
            "avg_patients":   avg,
            "required_staff": required,
            "capacity":       capacity,
            "utilization_pct": util,
            "understaffed":   avg > capacity and required > 0,
            "overstaffed":    util < 50 and required > 1,
            "zero_demand":    avg == 0,
        })
    return pd.DataFrame(rows)


@router.get("/")
def workforce_analysis(
    filename:      str = Query(DEFAULT_FILE),
    service_level: int = Query(DEFAULT_SL, ge=1, le=20),
):
    """Full shift analysis: required staff, shift summary, alerts, cost estimates."""
    df, wages = _load_clean(filename)
    comp = _compute(df, service_level, wages)

    # ── Per-shift summary ─────────────────────────────────────────
    shift_summary = []
    for shift, grp in comp.groupby("shift"):
        peak      = int(grp["required_staff"].max())
        avg_pts   = float(grp["avg_patients"].mean().round(2))
        n_windows = len(grp)
        wage      = wages.get(shift, 45)

        # Cost scenarios
        cost_peak      = peak * wage                        # always staff at peak
        cost_optimized = int(grp["required_staff"].sum() * wage / n_windows * n_windows)
        overstaff_hrs  = int(peak * n_windows - grp["required_staff"].sum())
        savings        = cost_peak - int(grp["required_staff"].sum() / n_windows * wage) if n_windows else 0

        shift_summary.append({
            "shift":           shift,
            "label":           SHIFT_NAMES.get(shift, shift),
            "recommended_staff": peak,
            "avg_patients_per_hour": avg_pts,
            "time_windows":    n_windows,
            "wage_per_shift_eur": wage,
            "cost_fixed_peak_eur":  cost_peak,
            "cost_optimized_eur":   cost_optimized,
            "overstaffing_hours":   overstaff_hrs,
            "potential_savings_eur": max(0, cost_peak - cost_optimized),
        })

    # ── Alerts ────────────────────────────────────────────────────
    alerts = []
    for _, r in comp.iterrows():
        if r["understaffed"]:
            alerts.append({
                "type":    "UNDERSTAFFING",
                "severity": "HIGH",
                "time":    r["time_window"],
                "shift":   r["shift"],
                "message": f"Demand ({r['avg_patients']} pts/hr) exceeds staff capacity ({r['capacity']} pts/hr). Add 1 more staff.",
            })
        if r["zero_demand"]:
            alerts.append({
                "type":    "ZERO_DEMAND",
                "severity": "INFO",
                "time":    r["time_window"],
                "shift":   r["shift"],
                "message": f"No patients expected. Consider reducing staffing.",
            })

    # ── Total daily cost ─────────────────────────────────────────
    total_cost_peak = sum(s["cost_fixed_peak_eur"] for s in shift_summary)
    total_cost_opt  = sum(s["cost_optimized_eur"]  for s in shift_summary)

    return {
        "service_level":     service_level,
        "wages_eur":         wages,
        "dataset_info":      df_info(df),
        "per_time_window":   comp.to_dict(orient="records"),
        "shift_summary":     shift_summary,
        "alerts":            alerts,
        "cost_summary": {
            "total_daily_cost_peak_staffing_eur":     total_cost_peak,
            "total_daily_cost_optimized_eur":         total_cost_opt,
            "total_potential_savings_eur":            total_cost_peak - total_cost_opt,
            "monthly_savings_if_optimized_eur":       (total_cost_peak - total_cost_opt) * 30,
        },
        "interpretation": (
            f"Staffing at the peak level for all shifts costs EUR {total_cost_peak}/day. "
            f"Dynamic staffing saves up to EUR {total_cost_peak - total_cost_opt}/day "
            f"(EUR {(total_cost_peak - total_cost_opt)*30}/month)."
        ),
    }


@router.get("/optimize")
def optimize(
    filename:      str   = Query(DEFAULT_FILE),
    service_level: int   = Query(DEFAULT_SL, ge=1, le=20),
    budget_eur:    float = Query(300.0, description="Daily staffing budget in EUR"),
):
    """
    Optimization scenarios: minimum cost, maximum coverage, budget-constrained.
    Returns recommended staff per shift under each scenario.
    """
    df, wages = _load_clean(filename)
    comp = _compute(df, service_level, wages)

    scenarios = {}
    for shift, grp in comp.groupby("shift"):
        wage   = wages.get(shift, 45)
        peak   = int(grp["required_staff"].max())
        minimum = max(1, int(math.ceil(grp["avg_patients"].mean() / service_level)))
        budget_staff = min(peak, max(1, int(budget_eur / wage)))

        scenarios[shift] = {
            "label":                  SHIFT_NAMES.get(shift, shift),
            "minimum_coverage":       minimum,
            "minimum_cost_eur":       minimum * wage,
            "full_coverage_staff":    peak,
            "full_coverage_cost_eur": peak * wage,
            "budget_constrained_staff": budget_staff,
            "budget_cost_eur":        budget_staff * wage,
            "recommended":            peak,
            "recommendation_reason":  (
                f"Peak of {peak} staff ensures 100% coverage at all times for EUR {peak*wage}/shift."
            ),
        }

    return {
        "budget_eur":    budget_eur,
        "service_level": service_level,
        "scenarios":     scenarios,
    }


@router.get("/alerts")
def alerts(
    filename:      str = Query(DEFAULT_FILE),
    service_level: int = Query(DEFAULT_SL, ge=1, le=20),
    threshold_pct: float = Query(80.0, description="Utilization % above which understaffing alert fires"),
):
    """
    Understaffing and overstaffing alerts with severity levels.
    """
    df, wages = _load_clean(filename)
    comp = _compute(df, service_level, wages)

    under, over, zero = [], [], []
    for _, r in comp.iterrows():
        if r["utilization_pct"] >= threshold_pct and r["required_staff"] > 0:
            under.append({
                "severity": "HIGH" if r["utilization_pct"] >= 95 else "MEDIUM",
                "time":     r["time_window"], "shift": r["shift"],
                "avg_patients": r["avg_patients"], "required_staff": r["required_staff"],
                "utilization_pct": r["utilization_pct"],
                "action": "Add 1 additional staff member immediately.",
            })
        if r["utilization_pct"] < 50 and r["required_staff"] > 1:
            wage = wages.get(r["shift"], 45)
            over.append({
                "severity": "LOW",
                "time":     r["time_window"], "shift": r["shift"],
                "avg_patients": r["avg_patients"], "required_staff": r["required_staff"],
                "utilization_pct": r["utilization_pct"],
                "wasted_cost_eur": round((r["required_staff"] - 1) * wage / 8, 2),
                "action": "Consider reducing by 1 staff or redistributing.",
            })
        if r["zero_demand"]:
            zero.append({"time": r["time_window"], "shift": r["shift"]})

    return {
        "service_level":        service_level,
        "threshold_pct":        threshold_pct,
        "understaffing_alerts": under,
        "overstaffing_alerts":  over,
        "zero_demand_windows":  zero,
        "total_alerts":         len(under) + len(over) + len(zero),
        "summary": f"{len(under)} understaffing · {len(over)} overstaffing · {len(zero)} zero-demand windows.",
    }


@router.get("/heatmap")
def heatmap(
    filename:      str = Query(DEFAULT_FILE),
    service_level: int = Query(DEFAULT_SL, ge=1, le=20),
):
    """
    Hour-by-shift demand matrix for heatmap chart.
    Returns: labels, values matrix, required staff matrix.
    """
    df, wages = _load_clean(filename)
    comp = _compute(df, service_level, wages)

    time_labels  = comp["time_window"].tolist()
    shift_labels = sorted(comp["shift"].dropna().unique().tolist())

    # Demand matrix: rows = time windows, cols = shifts
    demand_matrix = []
    staff_matrix  = []
    for _, r in comp.iterrows():
        row_d, row_s = [], []
        for s in shift_labels:
            if r["shift"] == s:
                row_d.append(float(r["avg_patients"]))
                row_s.append(int(r["required_staff"]))
            else:
                row_d.append(None)
                row_s.append(None)
        demand_matrix.append(row_d)
        staff_matrix.append(row_s)

    # Flat demand curve (for line chart)
    demand_curve = [{"time": r["time_window"], "patients": r["avg_patients"],
                     "shift": r["shift"], "staff": r["required_staff"]} for _, r in comp.iterrows()]

    return {
        "time_labels":    time_labels,
        "shift_labels":   [SHIFT_NAMES.get(s, s) for s in shift_labels],
        "shift_keys":     shift_labels,
        "demand_matrix":  demand_matrix,
        "staff_matrix":   staff_matrix,
        "demand_curve":   demand_curve,
    }


@router.get("/raw")
def raw(filename: str = Query(DEFAULT_FILE)):
    df, _ = _load_clean(filename)
    return {"data": df.fillna("").to_dict(orient="records"), "info": df_info(df)}
